src/server/mod.rs

```diff
@@ -1664,6 +1666,189 @@ impl LanguageServer for LspServer {
                 }
                 Ok(None)
             }
+            Ok(LspServerCommand::CompositionInitialize) => {
+                let command_params: CompositionInitParams =
+                    match serde_json::value::from_value(
+                        params.arguments[0].clone(),
+                    ) {
+                        Ok(value) => value,
+                        Err(err) => {
+                            return Err(LspError::InternalError(
+                                format!("{:?}", err),
+                            )
+                            .into())
+                        }
+                    };
+
+                let new_text = match self.state.lock() {
+                    Ok(mut state) => {
+                        if state.composition.is_none() {
+                            let file = self.store.get_ast_file(
+                                &command_params.text_document.uri,
+                            )?;
```

Comment:
Note: this does mean that we have 2 copies of the file in state:
* one in the `LspServer.store`
* another in the `LspServer.state.composition.file`

It's non-ideal. But works for now. 🤷🏼

Reply:
no longer is code. not relevant.